pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -411,9 +412,13 @@ private CompletableFuture<Boolean> isTopicOperationAllowed(TopicName topicName,
     private CompletableFuture<Boolean> isTopicOperationAllowed(TopicName topicName, String subscriptionName,
                                                                TopicOperation operation) {
         if (service.isAuthorizationEnabled()) {
-            AuthenticationDataSource authData =
-                    new AuthenticationDataSubscription(getAuthenticationData(), subscriptionName);
-            return isTopicOperationAllowed(topicName, operation, authData);
+            AuthenticationDataSource authDataSource =
+                    new AuthenticationDataSubscription(authenticationData, subscriptionName);
+            AuthenticationDataSource originalAuthDataSource = null;
+            if (originalAuthData != null) {
+                originalAuthDataSource = new AuthenticationDataSubscription(originalAuthData, subscriptionName);
+            }
+            return isTopicOperationAllowed(topicName, operation, authDataSource, originalAuthDataSource);
```

Comment:
I don't get why should it be so complicated? For the modified code,
- if `originalAuthData` is not `null`, `originalAuthData` will be used to construct the `AuthenticationDataSubscription` object as the 4th argument that will be used in `isTopicOperationAllowed`.
- Otherwise, the `authenticationData` will be used to construct the `AuthenticationDataSubscription` object (`authDataSource`) as the 3rd argument that will be used in `isTopicOperationAllowed`.

I think it's equivalent to

```java
            return isTopicOperationAllowed(topicName, operation,
                    new AuthenticationDataSubscription(getAuthenticationData(), subscriptionName), null);
```

After applying my change, `testVerifyOriginalPrincipalWithAuthDataForwardedFromProxy` will fail, but that's because you assume `allowTopicOperationAsync` will accept a non-null 4th argument.